### ISSUE TITLE
Eigen2Solver Build Errror

### ISSUE THREAD

**Issue by EmDash00** [sofa-framework/sofa#1412]
I'm getting a build error on the master branch of SOFA.

Using Eigen v3.3.7

```
Building CXX object SofaKernel/SofaBase/SofaBaseMechanics/CMakeFiles/SofaBaseMechanics.dir/BarycentricMapping.cpp.o
In file included from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix.h:35,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenSparseMatrix.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.inl:25,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.cpp:23:
/home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h:348:11: error: ‘ProductBase’ was not declared in this scope
  348 |  : traits<ProductBase<SparseTimeDenseProduct_MT<Lhs,Rhs>, Lhs, Rhs> >
      |           ^~~~~~~~~~~
/home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h:348:67: error: wrong number of template arguments (3, should be 1)
  348 |  : traits<ProductBase<SparseTimeDenseProduct_MT<Lhs,Rhs>, Lhs, Rhs> >
      |                                                                   ^
In file included from /usr/include/eigen3/Eigen/Core:365,
                 from /usr/include/eigen3/Eigen/SparseCore:11,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaHelper/src/sofa/helper/StateMask.h:27,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaCore/src/sofa/core/behavior/BaseMechanicalState.h:31,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaCore/src/sofa/core/behavior/MultiMatrixAccessor.h:25,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix.h:30,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenSparseMatrix.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.inl:25,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.cpp:23:
/usr/include/eigen3/Eigen/src/Core/util/ForwardDeclarations.h:17:29: note: provided for ‘template<class T> struct Eigen::internal::traits’
   17 | template<typename T> struct traits;
      |                             ^~~~~~
In file included from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix.h:35,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenSparseMatrix.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.h:26,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaBaseMechanics/BarycentricMapping.inl:25,
                 from /home/emdash00/Apps/git-apps/sofa/src/SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.cpp:23:
/home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h:348:69: error: expected ‘{’ before ‘>’ token
  348 |  : traits<ProductBase<SparseTimeDenseProduct_MT<Lhs,Rhs>, Lhs, Rhs> >
      |                                                                     ^
/home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h:457:23: error: expected template-name before ‘<’ token
  457 |   : public ProductBase<SparseTimeDenseProduct_MT<Lhs,Rhs>, Lhs, Rhs>
      |                       ^
/home/emdash00/Apps/git-apps/sofa/src/SofaKernel/SofaBase/../modules/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h:457:23: error: expected ‘{’ before ‘<’ token
make[2]: *** [SofaKernel/SofaBase/SofaBaseMechanics/CMakeFiles/SofaBaseMechanics.dir/build.make:206: SofaKernel/SofaBase/SofaBaseMechanics/CMakeFiles/SofaBaseMechanics.dir/BarycentricMapping.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1325: SofaKernel/SofaBase/SofaBaseMechanics/CMakeFiles/SofaBaseMechanics.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

**Comment by jnbrunet** [comment]
Thank you for the report @EmDash00 

This should not happen. `EigenBaseSparseMatrix_MT.h` should not be included when using Eigen >= 3.3.

I'm not sure why it is included in your compilation. 

I'll investigate. Looking at your output, I presume you are on Linux?

**Comment by EmDash00** [comment]
Yes I'm on Linux. I'm running Ubuntu 19.10.
